src/bounds.rs

```diff
@@ -0,0 +1,224 @@
+//! Definitions of bisection bounds.
+
+use crate::toolchains::{
+    download_progress, parse_to_naive_date, Toolchain, NIGHTLY_SERVER, YYYY_MM_DD,
+};
+use crate::GitDate;
+use crate::Opts;
+use crate::{today, EPOCH_COMMIT};
+use anyhow::bail;
+use chrono::NaiveDate;
+use reqwest::blocking::Client;
+use std::io::Read;
+use std::str::FromStr;
+
+/// A bisection boundary.
+#[derive(Clone, Debug)]
+pub enum Bound {
+    Commit(String),
+    Date(GitDate),
+}
+
+impl FromStr for Bound {
+    type Err = std::convert::Infallible;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        parse_to_naive_date(s)
+            .map(Self::Date)
+            .or_else(|_| Ok(Self::Commit(s.to_string())))
+    }
+}
+
+impl Bound {
+    /// Returns the SHA of this boundary.
+    ///
+    /// For nightlies, this will fetch from the network.
+    pub fn sha(&self) -> anyhow::Result<String> {
+        match self {
+            Bound::Commit(commit) => Ok(commit.clone()),
+            Bound::Date(date) => {
+                let date_str = date.format(YYYY_MM_DD);
+                let url =
+                    format!("{NIGHTLY_SERVER}/{date_str}/channel-rust-nightly-git-commit-hash.txt");
+
+                eprintln!("fetching {url}");
+                let client = Client::new();
+                let name = format!("nightly manifest {date_str}");
+                let mut response = download_progress(&client, &name, &url)?;
+                let mut commit = String::new();
+                response.read_to_string(&mut commit)?;
+
+                eprintln!("converted {date_str} to {commit}");
+
+                Ok(commit)
+            }
+        }
+    }
+}
+
+/// The starting bisection bounds.
+pub enum Bounds {
+    /// Indicates to search backwards from the given date to find the start
+    /// date where the regression does not occur.
+    SearchNightlyBackwards { end: GitDate },
+    /// Search between two commits.
+    Commits { start: String, end: String },
```

Comment:
IIUC, these aren't actually commits -- or at least indirectly so. E.g., `origin/master` is valid. Maybe we can note that in the comment?